Created API.md Octofet for Arduino boards. (#1)
amperka/Octofet: API.md

# Octofet API

## `class Octofet`

Create an object of type `Octofet` to communicate with a particular [Octofet board](https://my.amperka.com/modules/octofet). One Octofet board provides 8 power switches also known as “channels”.

### `Octofet(uint8_t pinCS, uint8_t deviceCount = 1)`

Constructs a new Octofet board object that uses the default hardware SPI bus.

> - `pinCS`: the `CS` pin used to control the shift-register latch. It is marked `CS` on the Octofet board. Any GPIO pin of the controller may be used.

Reviewer: ```suggestion
- `pinCS`: the chip select (also known as slave select) pin used to control the shift-register latch. It is marked `CS` on the Octofet board. Any GPIO pin of the controller may be used.
```

More informative

Reviewer 2: Slave Select usually called `SS`.

Reviewer 2: We have a wrongly named pin on the board. It is called CS. The action that it performs in the system is not a "chip select" at all. I propose to pronounce it as "Communication Start." I think this is more consistent with the actions performed by him.

Reviewer: These are details. The main idea is to match arg names to labels on the board and the traditional names of SPI... OK, we can write “the chip select (also known as SS or Slave Select) pin” but don’t overwhelm a reader more than that. Perhaps, it would be better to just drop this note:

“the chip select pin” ← period

BTW, what’s wrong with the Chip Select term? Why should it be Communication Start? I see the latter for the first time.

Reviewer 2: `what’s wrong with the Chip Select term?` - we select not one chip but a chain of chips.
`Why should it be Communication Start?` - before the start of communication, we switch-on CS, and after the end, we switch-of CS. 
Why Сommunication Start? I could not come up with another decoding suitable for the abbreviation CS. Why exactly `CS`? Because `CS` is already printed on the boards. All is IMHO.

Reviewer: I’m against synthetic new terms not seen in other places. CS is widely known as a part of SPI concept. The fact it potentially selects multiple chips depends on the point of view. You can think of the daisy-chain as of a single device with 16, 24, etc channels. Hope this makes sense.

Author: Ok, used option from @nkrkv

- `deviceCount`: the number of Octofet boards connected in a daisy-chain. If omitted, defines a single Octofet board.

### `Octofet(uint8_t pinCS, uint8_t pinMOSI, uint8_t pinSCK, uint8_t deviceCount = 1)`

Constructs a new Octofet board object that uses a software SPI bus.

- `pinCS`: the `CS` pin used to control the shift-register latch. It is marked `CS` on the Octofet board. Any GPIO pin of the controller may be used.
- `pinMOSI`: the `MOSI` pin used to send data to the peripherals. It is marked `DI` on the Octofet board. Any GPIO pin of the controller may be used.
- `pinSCK`: the `SCK` pin used to clock pulses which synchronize data transmission generated by the master. It is marked `П` on the Octofet board. Any GPIO pin of the controller may be used.

> - `deviceCount`: the number of Octofet boards connected in a daisy-chain. If omitted, defines a single Octofet board.

Reviewer: ```suggestion
- `pinCS`: the chip select (also known as slave select) pin used to control the shift-register latch. It is marked `CS` on the Octofet board. Any GPIO pin of the controller may be used.
- `pinMOSI`: the MOSI (Master Out Slave In) pin used to send data to the peripherals. It is marked `DI` on the Octofet board. Any GPIO pin of the controller may be used.
- `pinSCK`: the clock pin used to clock pulses which synchronize data transmission generated by the master. It is marked `П` on the Octofet board. Any GPIO pin of the controller may be used.
- `deviceCount`: the number of Octofet boards connected in a daisy-chain. If omitted, defines a single Octofet board.
```

Author: I suggest to write a description of all the pins by analogy:

- `pinCS`: the CS (Chip Select) pin used to control the shift-register latch. It is marked `CS` on the Octofet board. Any GPIO pin of the controller may be used.
- `pinMOSI`: the MOSI (Master Out Slave In) pin used to send data to the peripherals. It is marked `DI` on the Octofet board. Any GPIO pin of the controller may be used.
- `pinSCK`: the SCK (Serial Clock) pin used to clock pulses which synchronize data transmission generated by the master. It is marked `П` on the Octofet board. Any GPIO pin of the controller may be used.

Reviewer: You don’t have to be a robot when writing documentation:

- `readFoo`: reads Foo
- `readBar`: reads Bar
- `readQux`: reads Qux

So I see no point in repeating CS-CS MOSI-MOSI SCK-SCK pin-pin pin-pin-pin… From another point of view, it’s not a big deal. Not the reason for another bike shedding.

Author: Ok

# Octofet API

## `class Octofet`

> Uses: class Spi2Parallel

Reviewer: What does it mean? To be more precise, why should it be interesting to a documentation reader?

Reviewer 2: We say to the user: Spi2Parallel.cpp/h must be present in the src catalog.
Spi2Parallel is a low-level library for all pseudo-SPI devices. It supports both hardware and software SPI and provides data exchange between the controller and the chain. It is a common part for Octofet, MobyDig(big 7-segment), and we can use it with QuadDisplay2 and AmperkaFet.

Reviewer: I don’t get it. I bought Octofet, I see the phrase «Uses: class Spi2Parallel»… So what? What should I do? Will it work out of the box?

- If yes, remove this phrase
- If no, tell me (as to a user) what and why should I do to get it working

Author: Ok, we will not say the user how the board communicates at a low-level, more exact, with which class.

> Create an object of type `Octofet` to communicate with a particular [Octofet board](https://my.amperka.com/modules/octofet). The one board Octofet includes of 8 power switches.

Reviewer: ```suggestion
Create an object of type `Octofet` to communicate with a particular [Octofet board](https://my.amperka.com/modules/octofet). One Octofet board provides 8 power switches also known as “channels.”
```

Author: Why do you use quotation marks `“”`, not `""`. In the description of the method `digitalWrite`, we used quotation marks `""`.

### `Octofet(uint8_t pinCS, uint8_t deviceCount = 1)`

> Constructs a new Octofet board object that uses the hardware SPI bus with arguments:

Reviewer: ```suggestion
Constructs a new Octofet board object that uses the default hardware SPI bus.
```

Author: Done

> - `pinCS`: pin Chip Select used for controls the shift-register latch. Any GPIO-pin can be used.

Reviewer: ```suggestion
- `pinCS`: the chip select pin used to control the shift-register latch. Any GPIO pin may be used.
```

Author: Done

> - `deviceCount`: quantity of octofets boards connected in one chain: logical output pin headers of one module to the inputs of another. If the `deviceCount` argument is omitted, it used only one board Octofet.

Reviewer: ```suggestion
- `deviceCount`: the number of Octofet boards connected in a daisy-chain. If omitted, defines a single Octofet board.
```

Author: Done

### `Octofet(uint8_t pinCS, uint8_t pinMOSI, uint8_t pinSCK, uint8_t deviceCount = 1)`

> Constructs a new Octofet board object that uses the software SPI bus with arguments:

Reviewer: ```suggestion
Constructs a new Octofet board object that uses a software SPI bus.
```

Author: Done

> - `pinCS`: pin Chip Select used for controls the shift-register latch. Any GPIO-pin can be used.

Reviewer: ```suggestion
- `pinCS`: the chip select pin used to control the shift-register latch. Any GPIO pin may be used.
```

Author: Done

- `pinMOSI`: pin MOSI for sending data to the peripherals. Any GPIO-pin can be used.

> - `pinSCK`: pin SCK the clock pulses which synchronize data transmission generated by the master. Any GPIO-pin can be used.

Reviewer: Please, reword to match the descriptions above. Also, explicitly denote how these pins are labeled on the board. They are DI and П, not MOSI and SCK

Author: Done

> - `deviceCount`: quantity of octofets boards connected in one chain: logical output pin headers of one module to the inputs of another. If the `deviceCount` argument is omitted, it used only one board Octofet.

Reviewer: ditto

Author: Done

> Difference between hardware and software SPI is speed. Software SPI it will run much slower than hardware SPI, but hardware SPI works only on specific pins: `MOSI`, `MISO`(not used in Octofet) and `SCK`. Use software SPI only when hardware SPI pins are used by other devices.

Reviewer: ```suggestion
Difference between hardware and software SPI is speed. Software SPI runs slower than hardware SPI but hardware SPI works only on specific board pins: `MOSI`, `MISO` (not used in Octofet) and `SCK`.
```

Author: Done

### `void begin()`

> Initializes the given interface, prepares the board for communication. All power switches are set to turn-off state.

Reviewer: ```suggestion
Initializes the given interface, prepares the board for communication. All power switches are set to the turned-off state.
```

Author: Done

> Important, call this method at least once, in your `setup()`.

Reviewer: ```suggestion
Call this method before interfacing with Octofet. For example, in your `setup()`.
```

Author: Done

> ### `void digitalWrite(uint8_t switch, bool value, uint8_t device = 0)`

Reviewer: You can’t use `switch` as an arg name because `switch` is a keyword in C++. Use `switchIndex` or another name.

Also, I strongly suggest using the term “channel” instead of “switch” to avoid ambiguity. A switch might be considered as a mechanical commutation device like a button switch or reed switch.

Author: Done

> Sets the state ("on" or "off") of one power switch. Function arguments:

Reviewer: ```suggestion
Sets the state ("on" or "off") of one power switch.
```

Author: Done

> - `channel`: ordinal number power switch on device. Range is from `0` to `7`. Use option value is `ALL` to selected and sets state of all power switches.

Reviewer: ```suggestion
- `channel`: the power switch index. Ranges from `0` to `7`.
```

Do not complicate things. `ALL` is useless, let the user choose `digitalWrite8` instead.

Author: Ok

> - `value`: states selected power switch. Values (turn-on or turn-off) may be switch is different data types: `HIGH` or `LOW` / `true` or `false` / `1` or `0`.

Reviewer: ```suggestion
- `value`: defines the desired switch state. Valid values: `HIGH` / `1` / `true` to turn on or `LOW` / `0` / `false` to turn off.
```

Author: 👍 Cool

> - `device`: ordinal number of device on the bus. Range is from `0` to `n - 1`, where `n` is last number device. Use option value is `ALL` to selected and sets state of all power switches. If the `device` argument is omitted, it used only one board Octofet.

Reviewer: ```suggestion
- `device`: the index of the affected Octrofet in the daisy-chain. Ranges from `0` to `n - 1`, where `n` is the number of Octofets in the chain. If omitted, targets Octofet nearest to the controller.
```

Author: Done

> ### `void digitalWriteDevice(uint8_t value, uint8_t device = 0)`

Reviewer: So, you suggest names:

```cpp
digitalWrite(..., device)
digitalWriteDevice(..., device)
```

Throw the context out of your head. What is the difference now? Both methods relate to a device somehow. So, the “Device” refinement suffix does not clearly communicate the difference. The main idea is writing $eight $channels at $once. Just say it. Suggest one of the following:

```
digitalWrite8   (like in OpenGL)
digitalWriteAllChannels   (verbose but straightforward)
digitalWriteAtOnce   (wordy but straightforward)
```

Author: 1. Chose `digitalWrite8`.
2. Changed method name `getSwitchStateDevice` to `getSwitchState8`.

> Sets state (turn-on or turn-off) for all power switches. Function arguments:

Reviewer: ```suggestion
Sets state (turn-on or turn-off) for all power switches at once.
```

Author: Done

> - `value`: 8-bit value for all of 8 power switches. One bit for one power switch: 1 is turn-on and 0 is turn-off.

Reviewer: ```suggestion
- `value`: 8-bit value for all of 8 power switches. One bit for one channel: 1 is turn-on and 0 is turn-off. For example: `0b10101010`.
```

Author: Done

- `device`: ordinal number of device on the bus. Range is from `0` to `n - 1`, where `n` is last number device. Use option value is `ALL` to selected and sets state of all devices. If the `device` argument is omitted, it used only one board Octofet.

> ### `bool getSwitchState(uint8_t channel, uint8_t device = 0)`

Reviewer: To be consistent, would it be better to name these:

```
getChannelState
getChannelState8
```

Author: Done

> Reads state (turn-on or turn-off) for one power switch of the device. Returns `true` or `false`. Function arguments:

Reviewer: ```suggestion
Returns the last set (i.e., current) state for one power switch of the device. Returns `true` if turned on or `false` otherwise.
```

Author: Done

- `channel`: ordinal number power switch on device. Range is from `0` to `7`.

> - `device`: ordinal number of device on the bus. Range is from `0` to `n - 1`, where `n` is last number device. If the `device` argument is omitted, it used only one board Octofet.

Reviewer: Make consistent with above

Author: Done

### `uint8_t getChannelState8(uint8_t device = 0)`

> Reads state (turn-on or turn-off) for all power switches of the device. Returns 8-bit value for all of 8 power switches. One bit for one power switch: 1 is turn-on and 0 is turn-off. Function argument `device`: ordinal number of device on the bus. Range is from `0` to `n - 1`, where `n` is last number device. If the `device` argument is omitted, it used only one board Octofet.

Reviewer: Make consistent

Author: Done

# Octofet API

## `class Octofet`

Uses: class Spi2Parallel

> Create an object of type `Octofet` to communicate with a particular [Octofet board](https://my.amperka.com/modules/octofet).

Reviewer 3: ```suggestion
Creates an object of type `Octofet` to communicate with a particular [Octofet board](https://my.amperka.com/modules/octofet).
```

Author: Done

> ### `Octofet(uint8_t pinCS, uint8_t devicesCount = 1, uint8_t* resetValues = NULL)`

Reviewer 4: The constructor takes a `resetValues` array, and the deviceWrite function takes an 8-bit `value`. I think uniformity is needed in prototypes.

Author: @AlexErAmp Argument `resetValues` in Constructor `Octofet` takes values of power switchers on all boards devices. 
And Argument `value` in function `deviceWrite` takes values of power switchers only one board device. Added this to the function description.

> The one board Octofet includes of 8 power switches. The following convention are used: `pin` is a one power switch and `port` is a single Octofet device.

Reviewer 3: ```suggestion
One Octofet board includes 8 power switches. The following conventions are used: `pin` is a single power switch and `port` is a single Octofet device.
```

Author: Done.

> ### `Octofet(uint8_t pinCS, uint8_t devicesCount = 1)`

Reviewer: Grammar:

```suggestion
### `Octofet(uint8_t pinCS, uint8_t deviceCount = 1)`
```

Change in other places as well

Author: Done

> - `pinCS`: pin controls to enabled and disabled device on bus. Any GPIO-pin can be used.

Reviewer 2: I think better is:
pin Chip Select used for controls the shift-register latch.

Author: 👍

> Constructs a new object of data bus hardware SPI with arguments:

Reviewer 3: ```suggestion
Constructs a new object of the data bus hardware SPI with arguments:
```

Author: Done

> Constructs a new object of the data bus hardware SPI with arguments:

Reviewer: What? Object of the data bus :question:

Author: Fix in both description: hardware and software SPI

> - `pinCS`: pin Chip Select used for controls the shift-register latch. Any GPIO-pin can be used.

Reviewer 3: ```suggestion
- `pinCS`: pin Chip Select is used to control the shift-register latch. Any GPIO-pin can be used.
```

Author: Done.

> - `devicesCount`: quantity of octets boards connected in one chain: logical output pin headers of one module to the inputs of another. If the `devicesCount` argument is omitted, it used only one board Octofet.

Reviewer 2: octets -> octofet ?

Author: Oops!

> - `devicesCount`: quantity of octofets boards connected in one chain: logical output pin headers of one module to the inputs of another. If the `devicesCount` argument is omitted, it used only one board Octofet.

Reviewer 3: ```suggestion
- `devicesCount`: quantity of the Octofet boards connected in a chain: logical output pin of one module is connected to the input of another. If the `devicesCount` argument is omitted, it controls only one Octofet board.
```

Author: Done.

> - `devicesCount`: quantity of the Octofet boards connected in a chain: logical output pin of one module is connected to the input of another. If the `devicesCount` argument is omitted, it controls only one Octofet board.

Reviewer 3: ```suggestion
- `devicesCount`: quantity of the Octofet boards connected in a chain: the logical output pin of one module is connected to the input of another. If the `devicesCount` argument is omitted, it controls only one Octofet board.
```

Author: Done

> - `devicesCount`: quantity of the Octofet boards connected in a chain: the logical output pin of one module is connected to the input of another. If the `devicesCount` argument is omitted, it controls only one Octofet board.

Reviewer: Just above you say “pin is a single power switch” and “port is a single Octofet device”. So… `pinCS` is a power switch used to control a latch. And logical power switches of one module are connected to an input of another :astonished: 

Sounds like a mess.

Author: Removed the description about `pin` and `port`.

### `Octofet(uint8_t pinCS, uint8_t pinMOSI, uint8_t pinSCK, uint8_t devicesCount = 1)`

> Constructs a new object of data bus software SPI with arguments:

Reviewer 3: ```suggestion
Constructs a new object of the data bus software SPI with arguments:
```

Author: Done

> - `pinCS`: pin Chip Select used for controls the shift-register latch. Any GPIO-pin can be used.

Reviewer 3: ```suggestion
- `pinCS`: pin Chip Select is used to control the shift-register latch. Any GPIO-pin can be used.
```

Author: Done.

> ### `Octofet(uint8_t pinCS, uint8_t pinMOSI, uint8_t pinSCK, uint8_t devicesCount = 1, uint8_t* resetValues = NULL)`

Reviewer 4: The argument list is too long. The user can easily mix up arguments.

Author: Yes, but to solve this problem, we make `API.md`.

> - `pinMOSI`: pin MOSI for sending data to the peripherals. Any GPIO-pin can be used.

Reviewer 3: ```suggestion
- `pinMOSI`: pin MOSI is used to send the data to the peripherals. Any GPIO-pin can be used.
```

Author: Done.

> - `pinSCK`: pin SCK the clock pulses which synchronize data transmission generated by the master. Any GPIO-pin can be used.

Reviewer 3: ```suggestion
- `pinSCK`: pin SCK is used for the clock pulses which synchronize data transmission generated by the master. Any GPIO-pin can be used.
```

Author: Done

> - `devicesCount`: quantity of octofets boards connected in one chain: logical output pin headers of one module to the inputs of another. If the `devicesCount` argument is omitted, it used only one board Octofet.

Reviewer 3: ```suggestion
- `devicesCount`: quantity of the Octofet boards connected in a chain: the logical output pin of one module is connected to the input of another. If the `devicesCount` argument is omitted, it controls only one Octofet board.
```

Author: Done.

> Difference between hardware and software SPI is speed. Software SPI it will run much slower than hardware SPI, but hardware SPI works only on specific pins: `MOSI`, `MISO`(not used in Octofet) and `SCK`. Use software SPI only when hardware SPI pins are used by other devices.

Reviewer 3: ```suggestion
The difference between the hardware and the software SPI is in transmission speed and availability. The software SPI will run much slower than the hardware SPI, but the hardware SPI works only with specific pins: `MOSI`, `MISO`(not used in Octofet) and `SCK`. Use the software SPI only when the hardware SPI pins are already used by other devices.
```

Author: Done.

> ### `void begin(const uint8_t* resetValues = NULL)`

Reviewer: Use `nullptr` in modern C++. Do not use `NULL`: it is not type checked strictly.

```suggestion
### `void begin(const uint8_t* resetValues = nullptr)`
```

Author: Now, `resetValues` not used.

> Initializes the given interface, prepares the board for communication. The argument `resetValues` states power switches of all devices by default as 8-bit value. One bit for one power switch: 1 is turn-on and 0 is turn-off. If `resetValues` argument is adjusted, then when calling method `reset()`, power switches are set by resetValues. If `resetValues` argument is omitted, power switches are set to turn-off state.

Reviewer 3: ```suggestion
Initializes the given interface and prepares the board for communication. The argument `resetValues` sets the values of power switches of all devices by default as 8-bit. One bit for each power switch: 1 is "on" and 0 is "off". If the `resetValues` argument is adjusted, then when the `reset()` method is called, the power switches are set by resetValues. If the `resetValues` argument is omitted, the power switches are set to "off" state.
```

Author: Done.

> Important, call this method at least once, in your `setup()`.

Reviewer 3: ```suggestion
Important: call this method at least once, in your `setup()`.
```

Author: Done

> Initializes the given interface and prepares the board for communication. The argument `resetValues` sets the values of power switches of all devices by default as 8-bit. One bit for each power switch: 1 is "on" and 0 is "off". If the `resetValues` argument is adjusted, then when the `reset()` method is called, the power switches are set by resetValues. If the `resetValues` argument is omitted, the power switches are set to "off" state.

Reviewer 3: ```suggestion
Initializes the given interface and prepares the board for communication. The argument `resetValues` sets the values of power switches of all devices by default as 8-bit. One bit for each power switch: `1` is "on" and `0` is "off". If the `resetValues` argument is adjusted, then when the `reset()` method is called, the power switches are set by the `resetValues`. If the `resetValues` argument is omitted, the power switches are set to "off" state.
```

Author: Done

### `void digitalPinWrite(uint8_t pin, bool value, uint8_t device = 0)`

> Sets state (turn-on or turn-off) for one power switch. Function arguments:

Reviewer 3: ```suggestion
Sets the state ("on" or "off") of one power switch. Function arguments:
```

Author: I suggest change quotes `""` to `«»`. 

Sets the state («on» or «off») of one power switch. Function arguments:

> ### `digitalWrite(uint8_t device, uint8_t pin, bool value)`

Reviewer 4: If we make `device` the default argument and move it to the end of the list, we won’t need two `digitalWrite` functions.

Author: A nice idea.

> - `pin`: ordinal number power switch on device. Range is from `0` to `7`. Use option value is `ALL` to selected and sets state of all power switches.

Reviewer 3: ```suggestion
- `pin`: the ordinal number of the power switch on the board. It ranges from `0` to `7`. Use an optional value `ALL` to select and set the state of all power switches simultaneously.
```

Author: Done

> - `pin`: ordinal number power switch on device. Range is from `0` to `7`.

Reviewer 2: For device and pin, we have special value ALL = 255. You must say about it.

Author: Done.

> - `value`: states selected power switch. Values (turn-on or turn-off) may be switch is different data types: `HIGH` or `LOW` / `true` or `false` / `1` or `0`.

Reviewer 3: ```suggestion
- `value`: selects the state of the power switch. Values "on" and "off" may be put in different data types: `HIGH` or `LOW` / `true` or `false` / `1` or `0`.
```

Author: Done

> - `device`: ordinal number of device on the bus. Range is from `0` to `n - 1`, where `n` is last number device. Use option value is `ALL` to selected and sets state of all power switches. If the `device` argument is omitted, it used only one board Octofet.

Reviewer 3: ```suggestion
- `device`: the ordinal number of the device on the bus. It ranges from `0` to `n - 1`, where `n` is the last device number. Use an optional value `ALL` to select and set the state of all power switches simultaneously. If the `device` argument is omitted, it controls only one Octofet board.
```

Author: Done

### `void digitalPortWrite(uint8_t value, uint8_t device = 0)`

> Sets state (turn-on or turn-off) for all power switches. Function arguments:

Reviewer 3: ```suggestion
Sets the state ("on" or "off") for all power switches. Function arguments:
```

Author: Done

> - `value`: 8-bit value for all of 8 power switches. One bit for one power switch: 1 is turn-on and 0 is turn-off.

Reviewer 3: ```suggestion
- `value`: 8-bit value for all 8 power switches. One bit for each power switch: 1 is "on" and 0 is "off".
```

Author: Done

> - `value`: 8-bit value for all 8 power switches. One bit for each power switch: 1 is "on" and 0 is "off".

Reviewer 3: ```suggestion
- `value`: 8-bit value for all 8 power switches. One bit for each power switch: `1` is "on" and `0` is "off".
```

Author: Done

> ### `deviceWrite(uint8_t device, uint8_t value)`

Reviewer 4: We can make `device` the default argument, as with `digitalWrite`.

Author: A nice idea.

> - `device`: ordinal number of device on the bus. Range is from `0` to `n - 1`, where `n` is last number device. Use option value is `ALL` to selected and sets state of all devices. If the `device` argument is omitted, it used only one board Octofet.

Reviewer 3: ```suggestion
- `device`: the ordinal number of the device on the bus. It ranges from `0` to `n - 1`, where `n` is the last device number. Use an optional value `ALL` to select and set the state of all devices. If the `device` argument is omitted, it controls only one Octofet board.
```

Author: Done

### `bool digitalPinRead(uint8_t pin, uint8_t device = 0)`

> Reads state (turn-on or turn-off) for one power switch of the device. Returns `true` or `false`. Function arguments:

Reviewer 3: ```suggestion
Reads the state ("on" or "off") of a single power switch of the device. Returns `true` or `false`. Function arguments:
```

Author: Done

> - `device`: the ordinal number of the device on the bus. It ranges from `0` to `n - 1`, where `n` is the last device number. Use an optional value `ALL` to select and set the state of all devices. If the `device` argument is omitted, it controls only one Octofet board.

Reviewer: I find this pin vs port variation is confusing. I’d better stay away from custom terms and concepts invented for Octofet. What about:

```cpp
void digitalWrite(uint8_t pin, bool value, uint8_t device = 0);
void digitalWrite8(uint8_t state, uint8_t device = 0);
```

Author: I suggest these names of methods:

```cpp
void digitalWrite(uint8_t switch, bool value, uint8_t device = 0);
void digitalWriteDevice(uint8_t value, uint8_t device = 0);
```

> - `pin`: ordinal number power switch on device. Range is from `0` to `7`.

Reviewer 3: ```suggestion
- `pin`: the ordinal number power switch on the device. It ranges from `0` to `7`.
```

Author: Done

> ### `deviceWrite(uint8_t value)`

Reviewer 4: It might be useful to add a function `getDeviceValue` that returns the current values of the device.

Author: @AlexErAmp are you sure we need to return the power switchers states? Why is it for the user?Maybe create a function `toogle`, which inverts the current state power switch?

> - `device`: ordinal number of device on the bus. Range is from `0` to `n - 1`, where `n` is last number device. If the `device` argument is omitted, it used only one board Octofet.

Reviewer 3: ```suggestion
- `device`: the ordinal number of the device on the bus. It ranges from `0` to `n - 1`, where `n` is the last device number. If the `device` argument is omitted, it controls only one Octofet board.
```

Author: Done

Sets state for all power switches. The `value` argument is 8-bit value for all of 8 power switchers. One bit for one power switch: 1 is turn-on and 0 is turn-off.

A method is a special case of a function `deviceWrite(uint8_t device, uint8_t value)`, when there is only one device on the bus.

> ### `reset()`

Reviewer 2: **void** reset

> Reads state (turn-on or turn-off) for all power switches of the device. Returns 8-bit value for all of 8 power switches. One bit for one power switch: 1 is turn-on and 0 is turn-off. Function argument `device`: ordinal number of device on the bus. Range is from `0` to `n - 1`, where `n` is last number device. If the `device` argument is omitted, it used only one board Octofet.

Reviewer 3: ```suggestion
Reads state (turn-on or turn-off) of all power switches of the device. Returns the 8-bit value of all 8 power switches. One bit for each power switch: 1 is "on" and 0 is "off". Function argument `device`: the ordinal number of the device on the bus. It ranges from `0` to `n - 1`, where `n` is the last number device. If the `device` argument is omitted, it controls only one Octofet board.
```

Author: Done

> Resets values all power switches of all devices to default state.

Reviewer 3: ```suggestion
Resets values of all power switches of all devices to the default state.
```

Author: Done

> - If `resetValues` argument in constructs `Octofet` is adjusted, power switches are set by `resetValues`.

Reviewer 3: ```suggestion
- If the `resetValues` argument in the `Octofet` constructor is adjusted, power switches are set by the `resetValues`.
```

Author: Done

> Reads state (turn-on or turn-off) of all power switches of the device. Returns the 8-bit value of all 8 power switches. One bit for each power switch: 1 is "on" and 0 is "off". Function argument `device`: the ordinal number of the device on the bus. It ranges from `0` to `n - 1`, where `n` is the last number device. If the `device` argument is omitted, it controls only one Octofet board.

Reviewer 3: ```suggestion
Reads state (turn-on or turn-off) of all power switches of the device. Returns the 8-bit value of all 8 power switches. One bit for each power switch: `1` is "on" and `0` is "off". Function argument `device`: the ordinal number of the device on the bus. It ranges from `0` to `n - 1`, where `n` is the last number device. If the `device` argument is omitted, it controls only one Octofet board.
```

Author: Done

> Reads state (turn-on or turn-off) of all power switches of the device. Returns the 8-bit value of all 8 power switches. One bit for each power switch: `1` is "on" and `0` is "off". Function argument `device`: the ordinal number of the device on the bus. It ranges from `0` to `n - 1`, where `n` is the last number device. If the `device` argument is omitted, it controls only one Octofet board.

Reviewer: This creates an illusion of real reading. However, it just returns the internal state. I think this might be useful but should be named otherwise to avoid confusion. Like `getSwitchState` or something

Author: I suggest these names of methods:

```cpp
bool getSwitchState(uint8_t switch, uint8_t device = 0);
uint8_t getSwitchStateDevice(uint8_t device = 0);
```

> - If `resetValues` argument is omitted, power switches are set to turn-off state.

Reviewer 3: ```suggestion
- If the `resetValues` argument is omitted, power switches are set to the "off" state.
```

Author: Done

> - If the `resetValues` argument is omitted, power switches are set to the "off" state.

Reviewer: Why the `reset` concept is even necessary? I see how it makes things more complicated: more concepts, more arguments, more hidden state. And see no benefits.

Reviewer 2: Not all reset states are all zeros. For example "smart home" after reset:
The main light is off.
The standby light is on.
Heaters are off.
The refrigerator is on.
The lock is close. Etc.
The user will want to select the necessary state configuration for him after a reset. Yes, it can solve the problem using hardware "state inverters", but these devices increase complexity and reduce reliability.

Reviewer: Still does not make sense to me. Does the Octofet module stores defaults across reboots and power on/offs? If not, how this:

```cpp
uint8_t state = 0b10101010;
octofet.begin(&state);
```

is better than the following?

```cpp
octofet.begin();
octofet.digitalWrite8(0b10101010);
```

Reviewer 2: 1. Reset happens not only at the beginning of the program.
2. If state before is 0b1001 where the first bit is inductive device or device with a long reboot time then  0b1001 -> 0b0001 is more better then 0b1001 -> 0b0000 -> 0b0001.

Author: Method `reset` removed.

Call this method before interfacing with Octofet. For example, in your `setup()`.

### `void digitalWrite(uint8_t channel, bool value, uint8_t device = 0)`

Sets the state ("on" or "off") of one power switch.

- `channel`: the power switch index. Ranges from `0` to `7`.
- `value`: defines the desired switch state. Valid values: `HIGH` / `1` / `true` to turn on or `LOW` / `0` / `false` to turn off.
- `device`: the index of the affected Octofet in the daisy-chain. Ranges from `0` to `n - 1`, where `n` is the number of Octofets in the chain. If omitted, targets Octofet nearest to the controller.

### `void digitalWrite8(uint8_t value, uint8_t device = 0)`

Sets state (turn-on or turn-off) for all power switches at once.

- `value`: 8-bit value for all of 8 power switches. One bit for one channel: 1 is turn-on and 0 is turn-off. For example: `0b10101010`.
- `device`: the index of the affected Octofet in the daisy-chain. Ranges from `0` to `n - 1`, where `n` is the number of Octofets in the chain. If omitted, targets Octofet nearest to the controller.

### `bool getChannelState(uint8_t channel, uint8_t device = 0)`

> Returns the last set (i.e., current) state for one power switch of the device. Returns `true` if turned on or `false` otherwise.

Reviewer 3: ```suggestion
Returns the last (i.e., current) set state for one power switch of the device. Returns `true` if turned on or `false` otherwise.
```

Reviewer 3: 'cause last = current. Or it has to be "last set (i.e., current set)"

Reviewer: :ok_hand:

Author: Done

- `channel`: the power switch index. Ranges from `0` to `7`.
- `device`: the index of the affected Octofet in the daisy-chain. Ranges from `0` to `n - 1`, where `n` is the number of Octofets in the chain. If omitted, targets Octofet nearest to the controller.

### `uint8_t getChannelState8(uint8_t device = 0)`

> Returns 8-bit value of the last set (i.e., current) state for all 8 power switches of the device at once. One bit for one channel: 1 is turn-on and 0 is turn-off. For example: `0b10101010`. The argument `device` is the index of the affected Octofet in the daisy-chain. Ranges from `0` to `n - 1`, where `n` is the number of Octofets in the chain. If omitted, targets Octofet nearest to the controller.

Reviewer 3: ```suggestion
Returns 8-bit value of the last (i.e., current) set state for all 8 power switches of the device at once. One bit for one channel: 1 is turn-on and 0 is turn-off. For example: `0b10101010`. The argument `device` is the index of the affected Octofet in the daisy-chain. Ranges from `0` to `n - 1`, where `n` is the number of Octofets in the chain. If omitted, targets Octofet nearest to the controller.
```

Author: Done

Create an object of type `Octofet` to communicate with a particular [Octofet board](https://my.amperka.com/modules/octofet).

> The one board Octofet includes of 8 power switches. The following conversion are used: `pin` is a one power switch and `port` is one a device board Octofet.

Reviewer 4: `port` is one a device board Octofet. -> `port` is a single Octofet device.

### `Octofet(uint8_t pinCS, uint8_t devicesCount = 1)`

Constructs a new object of data bus hardware SPI with arguments:

- `pinCS`: pin Chip Select used for controls the shift-register latch. Any GPIO-pin can be used.
- `devicesCount`: quantity of octofets boards connected in one chain: logical output pin headers of one module to the inputs of another. If the `devicesCount` argument is omitted, it used only one board Octofet.

### `Octofet(uint8_t pinCS, uint8_t pinMOSI, uint8_t pinSCK, uint8_t devicesCount = 1)`

Constructs a new object of data bus software SPI with arguments:

- `pinCS`: pin Chip Select used for controls the shift-register latch. Any GPIO-pin can be used.
- `pinMOSI`: pin MOSI for sending data to the peripherals. Any GPIO-pin can be used.
- `pinSCK`: pin SCK the clock pulses which synchronize data transmission generated by the master. Any GPIO-pin can be used.
- `devicesCount`: quantity of octofets boards connected in one chain: logical output pin headers of one module to the inputs of another. If the `devicesCount` argument is omitted, it used only one board Octofet.

> Difference between hardware and software SPI is speed–with. Software SPI it will run much slower than hardware SPI, but hardware SPI works only on specific pins: `MOSI`, `MISO`(not used in Octofet) and `SCK`. Use software SPI only when hardware SPI pins are used by other devices.

Reviewer 4: speed–with -> speed.

### `void begin(uint8_t* resetValues = NULL)`

> Initializes the given interface, prepares the board for communication. The argument `resetValues` states power switches of all devices by default as 8-bit value. One bit for one power switch: 1 is turn-on and 0 is turn-off. If `resetValues` argument is adjusted, then when calling method `reset()`, power switches sets value by `resetValues`. If `resetValues` argument is omitted — power switches sets is turn-off.

Reviewer 4: power switches sets is turn-off. -> power switches are set to turn-off state.

- `pinSCK`: pin SCK the clock pulses which synchronize data transmission generated by the master. Any GPIO-pin can be used.
- `devicesCount`: quantity of octofets boards connected in one chain: logical output pin headers of one module to the inputs of another. If the `devicesCount` argument is omitted, it used only one board Octofet.

Difference between hardware and software SPI is speed. Software SPI it will run much slower than hardware SPI, but hardware SPI works only on specific pins: `MOSI`, `MISO`(not used in Octofet) and `SCK`. Use software SPI only when hardware SPI pins are used by other devices.

> ### `void begin(uint8_t* resetValues = NULL)`

Reviewer 4: const

> Initializes the given interface, prepares the board for communication. The argument `resetValues` states power switches of all devices by default as 8-bit value. One bit for one power switch: 1 is turn-on and 0 is turn-off. If `resetValues` argument is adjusted, then when calling method `reset()`, power switches sets value by `resetValues`. If `resetValues` argument is omitted — power switches are set to turn-off state.

Reviewer 4: 1) power switches sets value by `resetValues`. -> power switches are set by `resetValues`.
2) If `resetValues` argument is omitted — power switches are set to turn-off state.  ->  If `resetValues` argument is omitted, power switches are set to turn-off state.
